pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/DeletedObjectDao.java

```diff
@@ -0,0 +1,201 @@
+     * @param id         - id of the record we are deleting, primary key
+     * @param deleteType DeleteType
+     * @return
+     */
+    private static String getJsonRepOfDeletedObject(Connection conn, String id, DeleteType deleteType) {
+        //This method is to choose the right query, so we can avoid having dynamic queries with table names and risk SQL injection
+        String query;
+        if (deleteType.equals(DeleteType.ONC_HISTORY_DETAIL)) {
+            query = SQL_GET_JSON_OBJECT_FOR_ONC_HISTORY;
+        } else if (deleteType.equals(DeleteType.TISSUE)) {
+            query = SQL_GET_JSON_OBJECT_FOR_TISSUE;
+        } else if (deleteType.equals(DeleteType.SM_ID)) {
+            query = SQL_GET_JSON_OBJECT_FOR_SM_ID;
+        } else {
+            throw new DsmInternalError(String.format("DeleteType %s is not configured", deleteType));
+        }
+        return getJsonObject(conn, query, id, deleteType);
+    }
+
+    /**
+     * @param deleteType DeleteType
+     * @return the correct query for deleting from that table if the deleteType is from the alloweed tables, otherwise
+     * @throws DsmInternalError
+     */
+    private static String getDeleteQuery(DeleteType deleteType) {
+        //This method is to choose the right query, so we can avoid having dynamic queries with table names and risk SQL injection
+        String query = SQL_DELETE_VALUES;
+        if (deleteType.equals(DeleteType.ONC_HISTORY_DETAIL)) {
+            query = query.replace(TABLE_PLACE_HOLDER, DBConstants.DDP_ONC_HISTORY_DETAIL).replace(PK_PLACE_HOLDER, DBConstants.ONC_HISTORY_DETAIL_ID);
+        } else if (deleteType.equals(DeleteType.TISSUE)) {
+            query = query.replace(TABLE_PLACE_HOLDER, DBConstants.DDP_TISSUE).replace(PK_PLACE_HOLDER, DBConstants.TISSUE_ID);
+        } else if (deleteType.equals(DeleteType.SM_ID)) {
+            query = query.replace(TABLE_PLACE_HOLDER, DBConstants.SM_ID_TABLE).replace(PK_PLACE_HOLDER, DBConstants.SM_ID_PK);
+        } else {
+            throw new DsmInternalError(String.format("DeleteType %s is not configured", deleteType));
+        }
+        log.info("Created delete query for table {} with primary key {}");
+        return query;
+    }
+
+    /**
+     * returns a json blob of the object selected by id, the table to select from and the query to run
+     * is chosen based on the deleteType. The object can be used to be inserted in the deleted_object
+     * table later.
+     *
+     * @param conn       DB connection from the transaction that will do the whole delete procedure
+     * @param query      query to select from the right table
+     * @param id         primary key /id that is supposed to get deleted
+     * @param deleteType DeleteType of the patch.
+     * @return JSON representation of the object that will be deleted later
+     */
+
+    private static String getJsonObject(Connection conn, String query, String id, DeleteType deleteType) {
```

Comment:
Instead of using `@sql` in other methods and then querying the string value here via `SQL_GET_JSON_OBJECT_FOR...`, it might be simpler and more testable to do 

```
select
    column_name from information_schema.columns where table_name = 'sm_id'
 ```
and then concatenate everything on the java side so the java code makes a big select  json_object statement like:

```
select
    json_object('sm_id_value', sm_id_value, 'changed_by', changed_by)

from sm_id where sm_id_pk = 1
```

and then execute that query.

Reply:
That's a great approach! I did not see it that way